src/GMGPolar/solver.cpp

```diff
@@ -337,64 +280,72 @@ void GMGPolar::initializeSolution(const BoundaryConditions& boundary_conditions)
-            return true;
-        }
-    }
-    return false;
-}
+// =============================================================================
+//   Residual Handling Functions
+// =============================================================================
 
-std::pair<double, double> GMGPolar::computeExactError(Level& level, const Vector<double>& solution,
-                                                      Vector<double>& error)
+void GMGPolar::updateResidualNorms(Level& level, int iteration, double& initial_residual_norm,
+                                   double& current_residual_norm, double& current_relative_residual_norm)
 {
-    assert(exact_solution_ != nullptr);
+    level.computeResidual(level.residual(), level.rhs(), level.solution());
 
-    const PolarGrid& grid        = level.grid();
-    const LevelCache& levelCache = level.levelCache();
-    const auto& sin_theta_cache  = levelCache.sin_theta();
-    const auto& cos_theta_cache  = levelCache.cos_theta();
+    if (extrapolation_ != ExtrapolationType::NONE) {
+        Level& next_level = levels_[level.level_depth() + 1];
+        injection(level.level_depth(), next_level.solution(), level.solution());
+        next_level.computeResidual(next_level.residual(), next_level.rhs(), next_level.solution());
+        extrapolatedResidual(level.level_depth(), level.residual(), next_level.residual());
+    }
 
-    assert(solution.size() == error.size());
-    assert(solution.size() == grid.numberOfNodes());
+    switch (residual_norm_type_) {
+    case ResidualNormType::EUCLIDEAN:
+        current_residual_norm = l2_norm(level.residual());
+        break;
+    case ResidualNormType::WEIGHTED_EUCLIDEAN:
+        current_residual_norm = l2_norm(level.residual()) / std::sqrt(level.grid().numberOfNodes());
+        break;
+    case ResidualNormType::INFINITY_NORM:
+        current_residual_norm = infinity_norm(level.residual());
+        break;
+    default:
+        throw std::invalid_argument("Unknown ResidualNormType");
+    }
+    residual_norms_.push_back(current_residual_norm);
 
-#pragma omp parallel num_threads(threads_per_level_[level.level_depth()])
-    {
-#pragma omp for nowait
-        for (int i_r = 0; i_r < grid.numberSmootherCircles(); i_r++) {
-            double r = grid.radius(i_r);
-            for (int i_theta = 0; i_theta < grid.ntheta(); i_theta++) {
-                double theta                    = grid.theta(i_theta);
-                double sin_theta                = sin_theta_cache[i_theta];
-                double cos_theta                = cos_theta_cache[i_theta];
-                error[grid.index(i_r, i_theta)] = exact_solution_->exact_solution(r, theta, sin_theta, cos_theta) -
-                                                  solution[grid.index(i_r, i_theta)];
-            }
+    if (number_of_iterations_ == 0) {
+        if (!FMG_) {
+            initial_residual_norm = current_residual_norm;
```

Comment:
Is this variable badly named? If it represents the initial residual norm I'm not sure why you would need to update it during the solve?

Reply:
Found it :+1:

Seems a shame to have the `if (number_of_iterations_ == 0)` in every loop. Is that needed?